### PR TITLE
updated projectPath for VS Code AddOn

### PR DESCRIPTION
**Pull request by ifekonja**

### CODE REVIEW

**Comment by ifekonja** [comment]
The code doesnt work for svg files
I have made it by trick:
	let images_path= projectPath + 'images';
	for (let elementDefinition of elements) {
        if( elementDefinition.iconPath ) 
        {		
			let defaultContent= elementDefinition.defaultContent.replace(/images/gi, images_path);

Shall I integrate also this solution into pull request, or is there some other solution, because it always falls back to wrong url path

**Comment by jogibear9988** [comment]
maybe describe at first, what you are trying to achieve?

Do you have a set of widgets you'd like to use in different projects? Or are they in the project you edit?

Do you not use custom-elements? What is your project about?

**Comment by ifekonja** [comment]
1) Widget Library will be attached by extarnal TypeScript library (Angular)
2) I would like to work with "extends", in order to get derived classes. But it doesnt really work . One example for derived  user-code would be nice (like "toolButton").
3) I want make an editor , which manages HTML sites, whic can be merged together with angular code into runtime visualisation framework
4) I want to use my own icons/svg , so would be nice to set the external "asset" path (now I dont know, where it is actually determined (by "url.path ,or how?)
5) I want to  make my own mouse handler in order to capture events on custom "dvi"

**Comment by jogibear9988** [comment]
My plan for example is, to search the folder wich you opened in VSCode for "customelement.json" files and also the "node_modules" folder, and add them as elements in the designer addon. (https://github.com/webcomponents/custom-elements-manifest)

Additionally we could also think about another special file (designer specific, for wich we search in the projects), or some setting wich could be adjusted in Vscode settings, to link to external components

**Comment by ifekonja** [comment]
I will try to make something similar, which makes sense, since I have to work with multi-user projects , with class heirarchy, which is reflected to html-asset (so plain web server can load it over relative paths
Where do you load the "src"- referenced images?

**Comment by jogibear9988** [comment]
which images??? I have not yet used some

**Comment by ifekonja** [comment]
When I use "div" with "img" and "src=images/oneway.img" it loads from wrong URL. I am trying to figure that out where and adapt to some user path or at least take from href

// const path = url.pathname.replace("out/webview/designer.js", "");
let  filePath= original_url.href.replace("out/webview/designer.js","");

**Comment by jogibear9988** [comment]
yeah, you need to patch the paths (at least as I know) (maybe there is a way using the html base tag https://www.mediaevent.de/html/base.html)
look at the DesignerTextEditor.ts there I prepend all paths with a prefix via asWebViewUri. Maybe if you use this prefix in the urls.
This is nothing I've yet worked on

**Comment by jogibear9988** [comment]
Maybe try if this helps:

      <base href="${webview.asWebviewUri(vscode.workspace.workspaceFolders[0].uri)}" />

in designertexteditor head section of html

**Comment by ifekonja** [comment]
Another question: 
I want to extent the menu-panel with more functionality, like "SAVE", but it is very complicated.
I will make next "pull request" on this, since I dont see any other open option to do it.

serviceContainer.designerTools.set(NamedTools.Save, new SaveTool());

**Comment by ifekonja** [comment]
SOmething like 

addDesignerTool( "save", SaveTool, 0)   // position -> 0 ==begin, -1, end

**Comment by jogibear9988** [comment]
tools are extensions wich use pointer or keyboard events. And the tool list is a Set, so it has no order. 
So I don't see what you are trying to do.

If you like to create a ToolbarButton, that has not to be a tool, and as the list of designViewToolbarButtons is an arry, you can insert at any position as in every javascript array via splice.
So I again don't understand what you need.

**Comment by ifekonja** [comment]
I want to insert nice recognisable button (for SAVE) in the left ButtonPanel, which is useful.
I will make Pull Request on this

**Comment by jogibear9988** [comment]
maybe it is usefull, but I don't think this is something I want in standard designer, cause here it is thought only for tools.
Cause of that is extendable from outside, so if you need it in your project you could add your own buttons.

And for VSCode extension, it uses the save like every other TextDocument, so you can save via menu, via open documents, non of the has an additional save button

**Comment by ifekonja** [comment]
SAVE is actually "Save Configuration", where also .json file is being created (and verified against other project files), which is being used for other platform tasks.

**Comment by jogibear9988** [comment]
Yeah, means it‘s special for your usecase.

**Comment by ifekonja** [comment]
Is there any example of Adding the ToolButton  external Functionality?

**Comment by ifekonja** [comment]
The problem with relative path of "src='image/icon.svg'"  is located here in DefaultInstanceService.ts
   if (definition.defaultContent) {
      if (typeof definition.defaultContent === "string") {
        element.innerHTML = definition.defaultContent

setting "base" in HTML doesnt help
any other idea?

**Comment by ifekonja** [comment]
I have succeed to set the path:
src\vscode\DesignerTextEditor.ts

head
           base href="${webview.asWebviewUri(vscode.workspace.workspaceFolders[0].uri)}" 

----------------
and in designer.ts:

function prepend(elem, host) {
  var host = document.querySelector(host);
  var elem = document.createElement(elem);
  host.insertBefore(elem, host.firstChild);
}

function setBaseHref() {
	var baseElement = document.querySelector('base'); //document.getElementsByTagName('base')[0];
	if( baseElement===undefined)
	{
		prepend('base', 'head');
		baseElement = document.querySelector('base');
	}
	var currentBaseHrefValue = baseElement.getAttribute('href');
	var baseHrefUrl = projectPath;
	baseElement.setAttribute("href", baseHrefUrl);
}

setBaseHref();

--------

I will update pull request by this code

File: src/webview/designer.ts [hunk 1]
```diff
@@ -20,18 +22,65 @@ await window.customElements.whenDefined("node-projects-designer-view");
 const designerView = <DesignerView>document.querySelector("node-projects-designer-view");
 const propertyGrid = <PropertyGrid>document.getElementById("propertyGrid");
 const paletteView = <PaletteView>document.getElementById("paletteView");
+
+function setBaseHref(projectPath: string) {
```
Comment (jogibear9988) [inline src/webview/designer.ts:26]: why is this needed, when it is already set in DesignerTextEditor?

Comment (ifekonja) [reply src/webview/designer.ts:26]: To  change the "project" path dynamically

Comment (jogibear9988) [reply src/webview/designer.ts:26]: I don't understand what this is for. And how this could be usefull in default designer. I think this is especial for your usecase?

Comment (ifekonja) [reply src/webview/designer.ts:26]: You can decide it. If somebody wants to change project repository by clicking on tree explorer, it would be generally usefull

**Comment by ifekonja** [comment]
I have modified selection mechanism that way, it selects "custom-" div , when clicking on sub-child-div. A couple of lines ... Should I make pull request ... DesignItem.ts

**Comment by jogibear9988** [comment]
> I have modified selection mechanism that way, it selects "custom-" div , when clicking on sub-child-div. A couple of lines ... Should I make pull request ... DesignItem.ts

create a pull req, and I look. I will not include code wich is custom for your components, but if neccesarry then create a plugin point if needed. I'll take a look

File: src/webview/designer.ts (hunk 1 above)
Comment (jogibear9988) [reply src/webview/designer.ts:26]: I don't see where this pull request changes anything when clicked in tree explorer

**Comment by ifekonja** [comment]
OK, The goal is to get "custom-" divs to be selected as complex object , not allowing decoration div-s to be selected (and moved) at all, because main event handler should take care of that( since it is a complex action)

**Comment by jogibear9988** [comment]
> OK, The goal is to get "custom-" divs to be selected as complex object , not allowing decoration div-s to be selected (and moved) at all, because main event handler should take care of that( since it is a complex action)

Okay, show me your changes, and I look how we make interface you can use

**Comment by ifekonja** [comment]
designer.ts::: treeView comes later,, not yet so far...

**Comment by jogibear9988** [comment]
> designer.ts::: treeView comes later,, not yet so far...

yes, lack of time ...
It's only a hobby project

**Comment by jogibear9988** [comment]
basepath is now set in extension
